### ISSUE TITLE
Decide and document behavior of `map!` and `broadcast!` when function throws

### ISSUE THREAD

**Issue by chethega** [JuliaLang/julia#32047]
Currently, `broadcast!` into `BitArray` caches some results and only writes them back later. If the function that is broadcast happens to throw, then we leave the destination in an undefined state.

I think this is fine, and we should decide and document that `map!`, `broadcast!`, etc leave the destination with undefined contents if the mapped / broadcast function happens to throw. If we tried to make any guarantees we would prevent a lot of possible optimizations (e.g. hypothetical future multi-threaded implementations, caching without exception handlers, etc).

**Comment by vchuravy** [comment]
I think we should document that side-effect may happen in any order. We might even want to state that they may happen in parallel.

**Comment by mbauman** [comment]
Agreed.  We're already doing this with BitArrays and other packages are taking it even further with GPU kernels and such.   We do need to document it, though, and I agree we should go beyond errors: side-effects will happen, but their order is not guaranteed.  A good example of behavior I want to preserve is something like:

```julia
A .= rand.()
```

That should initialize all elements of `A` to be drawn from a random distribution, but the order in which they pull from the RNG is arbitrary.
